pkg/agent/openflow/network_policy.go

```diff
@@ -1313,3 +1335,69 @@ func (c *client) ReassignFlowPriorities(updates map[uint16]uint16, table binding
 	}
 	return nil
 }
+
+func parseDropFlow(flow string) (uint32, types.RuleMetric) {
+	// example format:
+	// table=101, n_packets=9, n_bytes=666, priority=200,ip,reg0=0x100000/0x100000,reg3=0x5 actions=drop
+	segs := strings.Split(flow, ",")
+	m := types.RuleMetric{}
+	pkts, _ := strconv.ParseUint(segs[1][strings.Index(segs[1], "=")+1:], 10, 32)
+	m.Packets = pkts
+	m.Sessions = pkts
+	bytes, _ := strconv.ParseUint(segs[2][strings.Index(segs[2], "=")+1:], 10, 32)
+	m.Bytes = bytes
+	id, _ := strconv.ParseUint(segs[6][strings.Index(segs[6], "0x")+2:strings.Index(segs[6], " ")], 16, 32)
+	return uint32(id), m
+}
+
+func parseAllowFlow(flow string) (uint32, types.RuleMetric) {
+	// example format:
+	// table=101, n_packets=0, n_bytes=0, priority=200,ct_state=-new,ct_label=0x1/0xffffffff,ip actions=goto_table:105
+	segs := strings.Split(flow, ",")
+	m := types.RuleMetric{}
+	pkts, _ := strconv.ParseUint(segs[1][strings.Index(segs[1], "=")+1:], 10, 32)
+	m.Packets = pkts
+	if strings.Contains(segs[4], "+") { // ct_state=+new
+		m.Sessions = pkts
+	}
+	bytes, _ := strconv.ParseUint(segs[2][strings.Index(segs[2], "=")+1:], 10, 32)
+	m.Bytes = bytes
+	idRaw := segs[5][strings.Index(segs[5], "0x")+2 : strings.Index(segs[5], "/")]
+	if len(idRaw) > 8 { // only 32 bits are valid.
+		idRaw = idRaw[:len(idRaw)-8]
+	}
+	id, _ := strconv.ParseUint(idRaw, 16, 32)
+	return uint32(id), m
+}
+
+func parseMetricFlow(flow string) (uint32, types.RuleMetric) {
+	dropIdentifier := "reg0"
+	if strings.Contains(flow, dropIdentifier) {
+		return parseDropFlow(flow)
+	} else {
+		return parseAllowFlow(flow)
+	}
+}
+
+func (c *client) NetworkPolicyMetrics() map[uint32]*types.RuleMetric {
+	result := map[uint32]*types.RuleMetric{}
+	egressFlows, _ := c.ovsctlClient.DumpTableFlows(uint8(EgressMetricTable))
+	ingressFlows, _ := c.ovsctlClient.DumpTableFlows(uint8(IngressMetricTable))
+	for _, flows := range [][]string{egressFlows, ingressFlows} {
+		for _, flow := range flows {
+			if strings.Contains(flow, "priority=0") {
+				continue
+			}
+			ruleID, metric := parseMetricFlow(flow)
+			// We have two flows for each allow rule. One with ct_state=+new matching calculates session numbers,
+			// and first packets, another ct_state=-new flow is used to calculate following packets. We need to merge
+			// metrics of these two flows to get the right number.
+			if accMetric, ok := result[ruleID]; ok {
+				accMetric.Merge(metric)
+			} else {
+				result[ruleID] = &metric
+			}
+		}
+	}
```

Comment:
I actually thought you were going to have something like this:
```go
collectMetricsFromFlows := func(flows []string) {
    for _, flow := range flows {
        // ...
    }
}

collectMetricsFromFlows(egressFlows)
collectMetricsFromFlows(ingressFlows)
```

(not a big fan of the `[][]string` and double for loop)

Reply:
Sure, updated.